numpy/core/arrayprint.py

```diff
@@ -634,6 +634,8 @@ def array2string(a, max_line_width=None, precision=None,
         msg = "array2string() got unexpected keyword argument '{}'"
         raise TypeError(msg.format(kwarg.popitem()[0]))
 
+    a = np.asanyarray(a)
```

Comment:
Note 1: This was needed because it turns out masked-arrays call `np.array2string(np.bool_(False))` when there is no mask, which technically wasn't allowed since `array2string` only accepts ndarrays. The fact that it worked before is a bit of an accident, it was failing now because `bool_` isn't a subclass of ndarray so the `np.ndarray.__getitem__` call below failed. Arguably the MA code is wrong here, but its just as easy to fix array2string to accept array-likes, as I've done here.

Note 2: This call can be replaced by `np.asarray` and tests still pass. If we do that, we don't need the `np.ndarray.__getitem__` call in `_formatArray` below. Does it make sense to drop the subclass at this point? It almost seems like it might... it would also avoid all the issues we had with matrices in this code.

Reply:
Yes, `asarray` may not be crazy for this routine. In a way cleaner than all the work-around. Just not sure what more would break...

Reply:
The alternative,would be to not have this ugly hack in the first place... And then just ducktype. But probably too late for that.

Reply:
It looks like we already discussed using `asarray` here two months ago: https://github.com/numpy/numpy/pull/10123#issuecomment-348548822 and we didn't have any objections then.

Reply:
Definitely don't see my comment above as objecting to it!

Reply:
Well understood, I just wanted to link in past discussions to so we don't just forget them.

Reply:
Should we update the docstring to say the argument is an array_like?

Reply:
An argument against this change - it makes this not work on duck types

Reply:
I suspect duck types of ndarray are hard to get working, so very rare. Also, the docstring specified it needs to be an ndarray.

Reply:
Do we actually need this here for the bug fix?

Reply:
Yeah, it can be removed now.